### PR TITLE
Implement `bal tool` command

### PR DESCRIPTION
**Pull request by gayaldassanayake**
## Purpose
$subject

Partially addresses https://github.com/ballerina-platform/ballerina-lang/issues/40264

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    private static final String TOOL_PULL_USAGE_TEXT = "bal tool pull <tool-id>[:<version>]";
+    private static final String TOOL_LIST_USAGE_TEXT = "bal tool list";
+    private static final String TOOL_REMOVE_USAGE_TEXT = "bal tool remove <tool-id>:[<version>]";
+    private static final String TOOL_SEARCH_USAGE_TEXT = "bal tool search [<tool-id>|<org>|<package>|<text>]";
+
+    private final boolean exitWhenFinish;
+    private final PrintStream outStream;
+    private final PrintStream errStream;
+
+    private final String distSpecificToolsTomlName = "dist-" + RepoUtils.getBallerinaShortVersion() + ".toml";
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:87]: Let's add these string values to class level constants atleast

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/CommandUtil.java [hunk 2]
```diff
@@ -942,4 +952,73 @@ private static String trimStartingWhitespaces(String str) {
...
+
+        // Delete package cache if available
+        Path packageCacheDir = cacheDir.resolve(orgName).resolve(packageName).resolve(version);
+        if (packageCacheDir.toFile().exists()) {
+            deleteDirectory(packageCacheDir);
+        }
+
+        // getResolution pulls all dependencies of the pulled package
+        PackageCompilation packageCompilation = balaProject.currentPackage().getCompilation();
+        Collection<Diagnostic> resolutionDiagnostics = packageCompilation.getResolution()
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/CommandUtil.java:1002]: we can break the logic from this line onwards to a separate method for readability.
`packageCompilation` as an argument and method will have the purpose of checking whether there are any error diagnostics. WDYT?

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-list.help [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+       bal tool list
+
+OPTIONS
+       -h, --help
+           Print the usage details of tool list command.
+
+DESCRIPTION
+       List the tool-id and the version of all locally available tools.
+
+       When there are multiple versions of the same tool, a row is printed
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-list.help:14]:
```suggestion
When multiple versions of the same tool exist, all available versions will be listed
```

Comment (Dilhasha) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-list.help:14]:
```suggestion
       When multiple versions of the same tool exist, all available versions will be listed
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+OPTIONS
+       -h, --help
+           Print the usage details of tool pull command.
+
+DESCRIPTION
+       Fetch a given bal tool from Ballerina Central repository and install it
+       as a command in the bal tool chain.
+
+       If the tool version is specified, that specific version will be fetched.
+       If the version is not specified, the latest version of the tool will
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help:16]:
```suggestion
       If the tool version is not specified, the latest version will be fetched.
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+       get pulled.
+
+       When there are multiple versions of the same tool pulled, the latest
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help:19]:
```suggestion
       When there are multiple versions of the same tool available locally, the latest
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-remove.help [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+           Print the usage details of tool remove command.
+
+DESCRIPTION
+       Remove an already installed tool from the bal tool chain.
+
+       If the tool version is specified, that specific version will be removed.
+       If the version is not specified, all versions of the tool will be
+       removed.
+
+       When the latest version of the tool is removed, the next latest version
```
Comment (Dilhasha) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-remove.help:18]:
```suggestion
       If the most recent version of the tool is removed, the next most recent version 
```

File: cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java [hunk 7]
```diff
@@ -566,6 +573,164 @@ public void pullPackage(String org, String name, String version, Path packagePat
...
+     */
+    public String[] pullTool(String toolId, String version, Path balaCacheDirPath, String supportedPlatform,
+                         String ballerinaVersion, boolean isBuild) throws CentralClientException {
+        String resourceUrl = "/" + TOOLS + "/" + toolId;
+        boolean enableOutputStream =
+                Boolean.parseBoolean(System.getProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM));
+        String toolSignature = toolId;
+
+        // TODO: remove once mocking is done
+        String url = "https://09edc8a7-fa97-48d7-b07a-b7709ce6101d.mock.pstmn.io" + resourceUrl;
```
Comment (Dilhasha) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:596]: @gayaldassanayake What's the purpose of this?

File: cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java [hunk 8]
```diff
@@ -772,6 +937,71 @@ public PackageSearchResult searchPackage(String query, String supportedPlatform,
         }
     }
 
+    /**
+     * Search packages in registry.
```
Comment (Dilhasha) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:941]:
```suggestion
     * Search tools in central.
```

File: cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java (hunk 7 above)
Comment (gayaldassanayake) [reply cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:596]: We haven't recieved the central APIs. So I'm using a mock URL. Need to replace with the central URL once recieved.

Comment (Dilhasha) [reply cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:596]: Ack

**Review by Dilhasha — APPROVED** [review]
We'll fix the suggestions in another PR.

**Pull request merged**